front/components/vaults/CreateOrEditVaultModal.tsx

```diff
@@ -168,59 +171,45 @@ export function CreateOrEditVaultModal({
     ];
   }, [selectedMembers]);
 
-  const createOrUpdateVault = async () => {
-    setIsSaving(true);
-    try {
-      if (selectedMembers.length > 0 && vault) {
-        const groupsUrl = `/api/w/${owner.sId}/groups/${vault.groupIds[0]}`;
-        const groupsRes = await fetch(groupsUrl, {
-          method: "PATCH",
-          headers: {
-            "Content-Type": "application/json",
-          },
-          body: JSON.stringify({
-            memberIds: selectedMembers,
-          }),
-        });
-
-        if (!groupsRes.ok) {
-          const groupsErrorData = await groupsRes.json();
-          throw new Error(
-            groupsErrorData.error?.message || "Failed to update vault members"
-          );
-        }
-      } else if (!vault) {
-        const url = `/api/w/${owner.sId}/vaults`;
-        const res = await fetch(url, {
-          method: "POST",
-          headers: {
-            "Content-Type": "application/json",
-          },
-          body: JSON.stringify({
-            name: vaultName,
-            memberIds: selectedMembers,
-          }),
-        });
+  const rows = useMemo(() => getTableRows(members), [members]);
 
-        if (!res.ok) {
-          const errorData = await res.json();
-          throw new Error(errorData.error?.message || "Failed to create vault");
-        }
-        const r = await res.json();
+  const columns = useMemo(() => getTableColumns(), [getTableColumns]);
 
-        await mutateVaults();
-        await mutateVaultsAsAdmin();
+  const onSave = useCallback(async () => {
+    setIsSaving(true);
 
-        await router.push(`/w/${owner.sId}/vaults/${r.vault.sId}`);
+    if (selectedMembers.length > 0 && vault) {
+      await doUpdate(vault, selectedMembers);
+    } else if (!vault) {
+      const createdVault = await doCreate(vaultName, selectedMembers);
+      if (createdVault) {
+        await router.push(`/w/${owner.sId}/vaults/${createdVault.sId}`);
       }
-    } finally {
-      setIsSaving(false);
     }
-  };
+    setIsSaving(false);
+    onClose();
+  }, [
+    doCreate,
+    doUpdate,
+    onClose,
+    owner.sId,
+    router,
+    selectedMembers,
+    vault,
+    vaultName,
+  ]);
 
-  const rows = useMemo(() => getTableRows(members), [members]);
+  const onDelete = useCallback(async () => {
+    if (!vault) {
+      return;
+    }
 
-  const columns = useMemo(() => getTableColumns(), [getTableColumns]);
+    const res = await doDelete(vault);
```

Comment:
I'm wondering if this `doDelete` shouldn't be in the modal itself, WDYT?

Reply:
let's revisit in another PR if you don't mind.